### ISSUE TITLE
[Submission] 45 params - Tom Bukic

### ISSUE THREAD

**Issue by tbukic** [anadim/AdderBoard#77]
### Author

tbukic

### Unique Parameter Count

45

### Accuracy

100.00% on 10,010 tests (verify.py seed=2025: 10K random + 10 edge cases)

### Method

Trained (SGD/Adam)

### Architecture

1L decoder-only transformer + circular arc embedding, d=3, 1h/1kv, hd=4, ff=2, RoPE theta=3, SwiGLU, RMSNorm

### Key Tricks

- Circular arc embedding (3 params instead of 30): digit tokens mapped via A*cos(start + stride*i)
- K = rotation(Q) (1-param 2D rotation in head_dim space replaces 12-param key projection)
- V = Q (value shares query weights exactly, 0 extra params — replaces 12-param value projection)
- Tied O=Q^T (output projection = transpose of query projection, saves 12 params)
- All 3 RMSNorms shared (ln1, ln2, final_norm share weights, saves 6 params)
- Tied lm_head to dynamic embedding table
- RoPE positional encoding (zero trainable params, theta=3)
- QK norms (RMSNorm on Q and K, stabilizes training)
- SwiGLU MLP at ff=2
- Curriculum learning + Grokfast-EMA gradient filter + adaptive weight decay
- Targeted fine-tuning

### Link to Code

https://github.com/tbukic/M10S-Transformer/blob/main/submissions/submission_45p.py

### Verification Output

```shell
Model: M10S-45p
Author: Tom Bukic
Parameters (unique): 45
Architecture: 1L decoder-only transformer + circular arc embed, d=3, 1h/1kv, hd=4, ff=2, K=rot(Q), V=Q, tieQO, shnorm

Results: 10010/10010 correct (100.00%)
Status: QUALIFIED (threshold: 99%)
```

### Additional Notes

- Parameter budget (45 total):
  - 3 circular arc embedding (A, start, stride)
  - 3 ln1 RMSNorm (shared with ln2 and final_norm via shnorm)
  - 1 k_rotation angle (K = rotation(Q), 2D rotation in head_dim=4 space)
  - 12 q_proj (shared with v_proj as V=Q, and o_proj via tieQO)
  - 4+4 QK norms (q_norm + k_norm)
  - 6 up_proj
  - 6 down_proj
  - 6 gate_proj
  Total: 3 + 3 + 1 + 12 + 8 + 6 + 6 + 6 = 45
- Training pipeline: K=rotation(Q) variant with curriculum + Grokfast-EMA + adaptive weight decay → targeted fine-tuning to push remaining errors to 0
- Extreme weight sharing: Q, V, and O all share the same 12-param projection (V=Q, O=Q^T), K uses a 1-param rotation of Q
- All 3 layer norms shared (single 3-param RMSNorm for ln1, ln2, final_norm)
